UnitystationLauncher/ContentScanning/AssemblyTypeChecker.cs

```diff
@@ -0,0 +1,1211 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Threading.Tasks;
+using ILVerify;
+using MoreLinq;
+using UnitystationLauncher.Services.Interface;
+
+// psst
+// You know ECMA-335 right? The specification for the CLI that .NET runs on?
+// Yeah, you need it to understand a lot of this code. So get a copy.
+// You know the cool thing?
+// ISO has a version that has correct PDF metadata so there's an actual table of contents.
+// Right here: https://standards.iso.org/ittf/PubliclyAvailableStandards/c058046_ISO_IEC_23271_2012(E).zip
+
+namespace UnitystationLauncher.ContentScanning;
+
+/// <summary>
+///     Manages the type white/black list of types and namespaces, and verifies assemblies against them.
+/// </summary>
+public sealed partial class AssemblyTypeChecker : IAssemblyChecker
+{
+    // Used to be in Sandbox.yml, moved out of there to facilitate faster loading.
+    private const string SystemAssemblyName = "mscorlib"; //TODO check security
+    //UnityEngine.dll
+    //mscorlib
+    //System.Runtime
+
+    /// <summary>
+    ///     Completely disables type checking, allowing everything.
+    /// </summary>
+    public bool DisableTypeCheck { get; init; }
+
+    public DumpFlags Dump { get; init; } = DumpFlags.None;
+    public bool VerifyIl { get; init; }
+
+    private readonly Task<SandboxConfig> _config;
+
+    private readonly IEnvironmentService _environmentService;
+
+    private readonly IFileService _fileService;
+
+    public AssemblyTypeChecker(IEnvironmentService environmentService, IFileService FileService)
+    {
+        _environmentService = environmentService;
+        VerifyIl = true;
+        DisableTypeCheck = false;
+        _fileService = FileService;
+        _config = Task.Run(LoadConfig);
+    }
+
+
+    private sealed class Resolver : IResolver
+    {
+        private readonly DirectoryInfo _managedPath;
+
+
+        private readonly Dictionary<string, PEReader> _dictionaryLookup = new Dictionary<string, PEReader>();
+
+        public void Dispose()
+        {
+            foreach (var Lookup in _dictionaryLookup)
+            {
+                Lookup.Value.Dispose();
+            }
+        }
+
+        PEReader IResolver.ResolveAssembly(AssemblyName assemblyName)
+        {
+            if (assemblyName.Name == null)
+            {
+                throw new FileNotFoundException("Unable to find " + assemblyName.FullName);
+            }
+
+            if (_dictionaryLookup.TryGetValue(assemblyName.Name, out var assembly))
+            {
+                return assembly;
+            }
+
+            FileInfo[] files = _managedPath.GetFiles("*.dll"); // Change the file extension to match your DLLs
```

Comment:
This might be missing some files in the scan on Linux/MacOS. I'm not sure the file extension MacOS uses, but Linux library files also use `.so` as the extension.

![image](https://github.com/unitystation/stationhub/assets/5573038/9230b58a-61ae-4aee-b588-9b91d3706c10)